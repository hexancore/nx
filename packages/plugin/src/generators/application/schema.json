{
  "$schema": "https://json-schema.org/schema",
  "$id": "Application",
  "title": "",
  "type": "object",
  "properties": {
    "name": {
      "type": "string",
      "description": "",
      "$default": {
        "$source": "argv",
        "index": 0
      },
      "x-prompt": "What name would you like to use?"
    },
    "directory": {
      "type": "string",
      "description": "A directory where the application is placed.",
      "x-priority": "important",
      "$default": {
        "$source": "argv",
        "index": 0
      },
    },
    "type": {
      "description": "The application type.",
      "type": "string",
      "enum": ["backend"],
      "x-prompt": "Which application type would you like to use ?",
      "x-priority": "important"
    }
  },
  "required": ["name", "type"]
}
