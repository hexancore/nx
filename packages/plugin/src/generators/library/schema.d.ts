import type { LibraryType } from "./LibraryType";

export interface LibraryGeneratorSchema {
  directory: string;
  type: LibraryType | 'backend' | 'frontend' | 'shared';
}