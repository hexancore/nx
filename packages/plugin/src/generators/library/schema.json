{
  "$schema": "https://json-schema.org/draft/2020-12/output/schema",
  "$id": "Library",
  "title": "",
  "type": "object",
  "properties": {
    "directory": {
      "type": "string",
      "description": "A directory where the library is placed in `apps/`. Project name will be directory with replaced '/'",
      "x-prompt": "Library project directory in `apps/`",
      "x-priority": "important",
      "$default": {
        "$source": "argv",
        "index": 0
      }
    },
    "type": {
      "description": "The library type.",
      "type": "string",
      "enum": ["backend", "shared", "frontend"],
      "x-prompt": "Which library type would you like to use ?",
      "x-priority": "important"
    }
  },
  "required": ["directory", "type"]
}
