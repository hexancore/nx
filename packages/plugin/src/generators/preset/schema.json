{
  "$schema": "https://json-schema.org/draft/2020-12/output/schema",
  "$id": "Preset",
  "title": "",
  "type": "object",
  "properties": {
    "name": {
      "type": "string",
      "description": "",
      "$default": {
        "$source": "argv",
        "index": 0
      },
      "x-prompt": "What name would you like to use?"
    }
  },
  "required": ["name"]
}
