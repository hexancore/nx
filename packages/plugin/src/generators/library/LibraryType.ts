export type LibraryType = 'frontend' | 'backend' | 'shared';