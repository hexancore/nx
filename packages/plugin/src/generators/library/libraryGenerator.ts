import {
  Tree,
  updateJson
} from '@nx/devkit';
import * as path from 'path';
import { HcNxHelper } from '../../util/HcNxHelper';
import { ProjectMeta, ProjectPackageJsonGenerator } from '../../util';
import { LibraryGeneratorSchema } from './schema';

export async function libraryGenerator(tree: Tree, options: LibraryGeneratorSchema): Promise<void> {

  const helper = new HcNxHelper(tree);
  const project: ProjectMeta = helper.createProjectMeta(options.directory, false);

  helper.addProjectConfiguration(project);
  ProjectPackageJsonGenerator.run(tree, { project });

  const filesPath = path.join(__dirname, 'files');
  helper.generateProjectFiles(project, path.join(filesPath, 'common'), '');

  updateJson(helper.tree, 'tsconfig.base.json', (v) => {
    v.compilerOptions.paths = v.paths ?? {};
    v.compilerOptions.paths[project.importName] = [`${project.root}/src/index.ts`];
    return v;
  });

  tree.write(path.join(project.root, 'src/index.ts'), 'export const INITIAL = 1;');
  tree.write(path.join(project.root, 'test/helper/.gitkeep'), '');
  tree.write(path.join(project.root, 'test/unit/.gitkeep'), '');
  tree.write(path.join(project.root, 'test/integration/.gitkeep'), '');
  helper.generateProjectFiles(project, path.join(filesPath, 'test'), 'test/unit');
}

export default libraryGenerator;
