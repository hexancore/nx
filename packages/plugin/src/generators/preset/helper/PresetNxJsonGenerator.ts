import {
  Tree,
  updateJson,
  type TargetConfiguration
} from '@nx/devkit';

export type ProjectTargetsConfiguration = Record<string, TargetConfiguration>;

export class PresetNxJsonGenerator {

  public static run(tree: Tree) {
    const g = new this();
    g.apply(tree);
  }

  public apply(tree: Tree): void {
    updateJson(tree, 'nx.json', (o) => {
      o.extends = 'nx/presets/npm.json';
      o.pluginsConfig = {
        "@nx/js": {
          "analyzeSourceFiles": true
        }
      };

      o.namedInputs = { ...(o.namedInputs ?? {}), ...this.namedInputs() };
      o.targetDefaults = { ...(o.targetDefaults ?? {}), ...this.targetDefaults() };
      o.generators = {
        "@nx/workspace:move": {
          "projectNameAndRootFormat": "as-provided"
        }
      };
      o.useInferencePlugins = false;
      o.defaultBase = 'main';

      o['workspaceLayout'] = {
        "appsDir": "apps",
        "libsDir": "libs"
      };

      return o;
    });
  }

  private namedInputs(): Record<string, any> {
    return {
      default: [
        "{projectRoot}/**/*",
        "sharedGlobals"
      ],
      sharedGlobals: [],
      production: [
        "default",
        "!{projectRoot}/test/**/*",
        "!{projectRoot}/jest.config.{js,ts,mjs,mts}",
        "!{projectRoot}/tsconfig.test.json",
        "!{projectRoot}/.eslintrc.json",
        "{workspaceRoot}/.workspace/vite/**/*"
      ],
      storybook: [
        "production",
        "{projectRoot}/**/*.stories.ts",
        "{workspaceRoot}/.workspace/storybook/**/*",
      ]
    };
  }

  private targetDefaults(): ProjectTargetsConfiguration {
    return {
      ...this.jsTargetsDefaults(),
      ...this.eslintTargetsDefaults(),
      ...this.jestTargetsDefaults(),
      ...this.viteTargetsDefaults(),
      ...this.storybookTargetsDefaults(),
    };
  }

  private jsTargetsDefaults(): ProjectTargetsConfiguration {
    return {
      "@nx/js:tsc": {
        cache: true,
        "inputs": ["production", "^production"],
        "outputs": ["{options.outputPath}"],
        "dependsOn": ["^build", "test"],
        "options": {
          "outputPath": "{workspaceRoot}/dist/{projectRoot}",
          "tsConfig": "{projectRoot}/tsconfig.build.json",
          "transformers": [
            {
              "name": "typescript-transform-paths/nx-transformer"
            }
          ]
        }
      }
    };
  }

  private eslintTargetsDefaults(): ProjectTargetsConfiguration {
    return {
      "@nx/eslint:lint": {
        "inputs": ["default", "{workspaceRoot}/.eslintrc.json", "{workspaceRoot}/.eslintignore"],
        "cache": true,
        "options": {
          "lintFilePatterns": ["{projectRoot}/**/*.ts", "{projectRoot}/package.json"]
        },
        "configurations": {
          "fix": {
            "fix": true,
            "cache": false
          }
        }
      },
    };
  }

  private jestTargetsDefaults(): ProjectTargetsConfiguration {
    return {
      "@nx/jest:jest": {
        "cache": true,
        "inputs": ["default", "{workspaceRoot}/.workspace/jest/**/*"],
        "options": {
          "jestConfig": "{projectRoot}/jest.config.ts",
          "runInBand": true,
          "passWithNoTests": true
        },
        "configurations": {
          "watch": {
            "watch": true
          },
          "ci": {
            "ci": true,
            "codeCoverage": true
          },
          "updateSnapshot": {
            "updateSnapshot": true
          },
          "codeCoverage": {
            codeCoverage: true,
          }
        }
      },
    };
  }

  private viteTargetsDefaults(): ProjectTargetsConfiguration {
    return {
      "@nx/vite:build": {
        cache: true,
        "inputs": ["production", "^production"],
        outputs: ["{options.outputPath}"],
        defaultConfiguration: "production",
        options: {
          "outputPath": "dist/{projectRoot}",
          "skipTypeCheck": true,
          "configFile": "{projectRoot}/vite.config.mts",
          "tsConfig": "{projectRoot}/tsconfig.build.json",
          "generatePackageJson": false
        },
        configurations: {
          development: {
            "mode": "development"
          },
          production: {
            "mode": "production"
          }
        }
      },
      "@nx/vite:preview-server": {
        defaultConfiguration: "development",
        cache: false,
        options: {
          "buildTarget": "build"
        },
        configurations: {
          development: {
            buildTarget: "build:development"
          },
          production: {
            buildTarget: "build:production"
          }
        }
      },
      "@nx/vite:dev-server": {
        cache: false,
        defaultConfiguration: "development",
        options: {
          buildTarget: "build"
        },
        configurations: {
          development: {
            buildTarget: "build:development"
          },
          production: {
            buildTarget: "build:production"
          }
        }
      },
      "@nx/vite:test": {
        cache: true,
        inputs: ["default"],
        options: {
          configFile: "{projectRoot}/vite.config.mts",
          passWithNoTests: true,
          reportsDirectory: "{workspaceRoot}/coverage/{projectRoot}"
        },
        configurations: {
          watch: {
            watch: true
          }
        }
      },
    };
  }

  private storybookTargetsDefaults(): ProjectTargetsConfiguration {
    return {
      "@nx/storybook:storybook": {
        "cache": false,
        "options": {
          "port": 4400,
          "configDir": "{projectRoot}/.storybook"
        },
        "configurations": {
          "ci": {
            "quiet": true
          }
        }
      },
      "@nx/storybook:build": {
        "inputs": ["storybook"],
        "cache": true,
        "outputs": ["{options.outputDir}"],
        "options": {
          "outputDir": "{workspaceRoot}/dist/storybook/{projectRoot}",
          "configDir": "{projectRoot}/.storybook"
        },
        "configurations": {
          "ci": {
            "quiet": true
          }
        }
      },
      "storybook-test": {
        "executor": "nx:run-commands",
        "options": {
          "command": "test-storybook -c {projectRoot}/.storybook --url=http://localhost:4400"
        }
      },
      "storybook-static": {
        "executor": "@nx/web:file-server",
        "options": {
          "buildTarget": "storybook-build",
          "staticFilePath": "{workspaceRoot}/dist/storybook/{projectRoot}",
          "spa": true
        },
        "configurations": {
          "ci": {
            "buildTarget": "storybook-build:ci"
          }
        }
      }
    };
  }
}