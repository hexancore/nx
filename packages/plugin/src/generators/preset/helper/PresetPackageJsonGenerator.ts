import {
  addDependenciesToPackageJson,
  updateJson,
  Tree,
} from '@nx/devkit';

export class PresetPackageJsonGenerator {

  public constructor(
    private nxVersion: string,
    private hcNxPluginVersion: string
  ) {

  }

  public static run(tree: Tree, options: {
    nxVersion: string,
    hcNxPluginVersion: string
  }) {
    const g = new this(options.nxVersion, options.hcNxPluginVersion);
    g.apply(tree);
  }

  public apply(tree: Tree): void {
    this.generateDeps(tree);

    updateJson(tree, 'package.json', (o) => {
      o.license = 'UNLICENSED';
      o.scripts = o.scripts ?? {};

      o.scripts['prepare'] = "node .husky/install.mjs",
        o.scripts['precommit'] = "nx run-many --targets=lint,test,build";
      o.scripts['ci'] = "nx run-many --targets=lint,test,build,e2e --configuration=ci";
      return o;
    });
  }

  private generateDeps(tree: Tree): void {
    const deps = this.createDeps();
    const devDeps = this.createDevDeps();
    addDependenciesToPackageJson(tree, deps, devDeps);
  }

  private createDeps(): Record<string, string> {
    const hexancore = {
      '@hexancore/core': '0.14.*',
      '@hexancore/common': '0.13.*',
      '@hexancore/cloud': '0.1.*',
      '@hexancore/auth': '0.3.*',
      '@hexancore/typeorm': '0.14.*',
    };

    return {
      ...hexancore,
      'tslib': '2.6.2'
    };
  }

  private createDevDeps(): Record<string, string> {
    return {
      ...this.createHexancoreDevDeps(),
      ...this.createNxDevDeps(),
      ...this.createFrontendDevDeps(),
      ...this.createJestDevDeps(),
      ...this.createEslintDevDeps(),
      ...this.createEssentialsDevDeps()
    };
  }

  private createNxDevDeps(): Record<string, string> {
    return {
      "@nx/devkit": this.nxVersion,
      "@nx/esbuild": this.nxVersion,
      "@nx/eslint-plugin": this.nxVersion,
      "@nx/jest": this.nxVersion,
      "@nx/js": this.nxVersion,
      "@nx/linter": this.nxVersion,
      "@nx/plugin": this.nxVersion,
    };
  }

  private createHexancoreDevDeps(): Record<string, string> {
    const [major, minor] = this.hcNxPluginVersion.split('.');

    return {
      "@hexancore/nx": `${major}.${minor}.*`,
      "@hexancore/mocker": "^1.1.2",
    };
  }

  private createFrontendDevDeps(): Record<string, string> {
    const vue = {
      "@vue/eslint-config-typescript": "^11.0.3",
      "@vue/test-utils": "^2.4.5",
      "@vue/tsconfig": "^0.4.0",
      "vue": "^3.4.25",
      "vue-eslint-parser": "^9.4.2",
      "vue-router": "^4.3.2",
      "pinia": "^2.1.7",
      "vue-tsc": "^1.8.27",
      "vee-validate": "^4.12.6",
    };

    const vueI18n = {
      "vue-i18n": "^9.13.1",
      "@intlify/eslint-plugin-vue-i18n": "^2.0.0",
      "@intlify/unplugin-vue-i18n": "^2.0.0",
      "unplugin-swc": "^1.4.5",
    };

    const vite = {
      "@vitejs/plugin-basic-ssl": "^1.1.0",
      "@vitejs/plugin-vue": "^5.0.4",
      "@vitest/coverage-v8": "1.6.*",
      "@vitest/ui": "1.6.*",
      "vitest": "1.6.*",
      "@rollup/plugin-node-resolve": "^15.2.3",
      "rollup": "^4.17.0",
      "vite": "^5.2.10",
      "vite-tsconfig-paths": "^4.3.2",
      "autoprefixer": "^10.4.19",
      "@babel/core": "^7.24.4",
      "@babel/plugin-transform-modules-commonjs": "^7.24.1",
      "@babel/preset-env": "^7.24.4",
    };

    const primevue = {
      "primeflex": "^3.3.1",
      "primeicons": "^6.0.1",
      "primevue": "^3.52.0",
    };

    const style = {
      "postcss": "^8.4.38",
      "sass": "^1.75.0",
      "stylelint": "^15.11.0",
      "stylelint-config-standard": "^33.0.0",
      "@samatech/postcss-basics": "^0.6.0",
    };

    return {
      ...vue,
      ...vueI18n,
      ...vite,
      ...primevue,
      ...style
    };
  }

  public createEslintDevDeps(): Record<string, string> {
    return {
      "eslint": "8.57.0",
      "eslint-plugin-import": "^2.29.1",
      "eslint-plugin-vue": "^9.25.0",
      "@typescript-eslint/parser": "7.3.0",
      "@stylistic/eslint-plugin": "^2.1.0",
    };
  }

  public createJestDevDeps(): Record<string, string> {
    return {
      "happy-dom": "^9.20.3",
      "jest": "^29.7.0",
      "jest-docblock": "^29.7.0",
      "jest-environment-jsdom": "^29.7.0",
      "jest-environment-node": "^29.7.0",
      "jest-expect-message": "^1.1.3",
      "jest-runner": "^29.7.0",
      "jest-runner-groups": "^2.2.0",
      "jsdom": "~22.1.0",
      "ts-jest": "^29.1.2",
      "@types/jest": "^29.5.12",
      "@jest/globals": "^29.7.0",
    };
  }

  public createEssentialsDevDeps(): Record<string, string> {
    return {
      "tsconfig": "^7.0.0",
      "typescript": "5.4.5",
      "typescript-transform-paths": "^3.4.7",
      "esbuild": "^0.19.12",
      "axios-mock-adapter": "^1.22.0",
      "@types/node": "20.12.12",
      "@nestjs/testing": "^10.3.8",
      "husky": "^9.0.11",
    };
  }
}