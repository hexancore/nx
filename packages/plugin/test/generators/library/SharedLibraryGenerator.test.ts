import { Tree } from '@nx/devkit';

import { libraryGenerator } from '../../../src/generators/library/libraryGenerator';
import { LibraryGeneratorSchema } from '../../../src/generators/library/schema';
import { createEmptyNxTree, exceptOnlyFilesFromListExistInWorkspace, exceptWorkspaceFileMatchSnapshot, expectedFiles } from '../../helper/functions';

describe('Library Generator: Shared', () => {
  let tree: Tree;
  const type = 'shared';
  const projectDirectory = `task-manager/${type}`;
  const projectRoot = `libs/${projectDirectory}`;
  beforeAll(async () => {
    tree = createEmptyNxTree();
    const options: LibraryGeneratorSchema = { directory: projectDirectory, type };
    await libraryGenerator(tree, options);
  });

  const expectedProjectFilesToMatchSnapshots = [
    'README.md',
    'project.json',
    'package.json',
    '.eslintrc.json',
    'jest.config.ts',
    'tsconfig.json',
    'tsconfig.test.json',
    'tsconfig.build.json',
  ].map(f => `${projectRoot}/` + f);

  const expectedWorkspaceFilesToMatchSnapshots = [
    'tsconfig.base.json',
    'nx.json',
    'package.json',
    ...expectedProjectFilesToMatchSnapshots
  ];

  test.each(expectedWorkspaceFilesToMatchSnapshots.map(i => [i]))('%s should match snapshot', (filePath) => {
    exceptWorkspaceFileMatchSnapshot(tree, filePath);
  });

  test('only file from list should exist', () => {
    const expectedProjectFiles = expectedFiles(`${projectRoot}/`, [
      'src/index.ts',
      'test/config.ts',
      'test/unit/sample.test.ts',
      'test/helper/.gitkeep',
      'test/integration/.gitkeep',
      'src/Service/.gitkeep'
    ]);
    const expected = [
      '.prettierrc',
      ...expectedProjectFiles,
      ...expectedWorkspaceFilesToMatchSnapshots
    ];
    exceptOnlyFilesFromListExistInWorkspace(tree, expected);
  });
});
