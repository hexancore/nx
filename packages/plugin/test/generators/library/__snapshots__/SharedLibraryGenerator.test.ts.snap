// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Library Generator: Shared libs/task-manager/shared/.eslintrc.json should match snapshot 1`] = `
"{
  "extends": ["../../../.eslintrc.json"],
  "ignorePatterns": ["!**/*"]

}
"
`;

exports[`Library Generator: Shared libs/task-manager/shared/README.md should match snapshot 1`] = `
"# task-manager-shared

"
`;

exports[`Library Generator: Shared libs/task-manager/shared/jest.config.ts should match snapshot 1`] = `
"import { defineProjectJestConfig } from "../../../.workspace/jest/jest.preset";

const project = {
  workspaceRootRelative: '../../..',
  root: 'libs/task-manager/shared',
  name: 'task-manager-shared'
};

export default defineProjectJestConfig(__dirname, project);"
`;

exports[`Library Generator: Shared libs/task-manager/shared/package.json should match snapshot 1`] = `
"{
  "name": "@testcompany/task-manager-shared",
  "type": "commonjs",
  "private": true,
  "dependencies": {
    "tslib": "2.6.2"
  },
  "main": "./src/index.js",
  "types": "./src/index.d.ts"
}
"
`;

exports[`Library Generator: Shared libs/task-manager/shared/project.json should match snapshot 1`] = `
"{
  "name": "task-manager-shared",
  "$schema": "../../../node_modules/nx/schemas/project-schema.json",
  "projectType": "library",
  "sourceRoot": "libs/task-manager/shared/src",
  "targets": {
    "lint": {
      "executor": "@nx/eslint:lint"
    },
    "test": {
      "executor": "@nx/jest:jest"
    },
    "build": {
      "executor": "@nx/js:tsc",
      "options": {
        "main": "{projectRoot}/src/index.ts"
      }
    }
  }
}
"
`;

exports[`Library Generator: Shared libs/task-manager/shared/tsconfig.build.json should match snapshot 1`] = `
"{
  "extends": "./tsconfig.json",
  "compilerOptions": {
    "outDir": "../../../dist/out-tsc",
    "declaration": true
  },
  "include": ["src/**/*.ts"],
  "exclude": ["test/**/*"]
}
"
`;

exports[`Library Generator: Shared libs/task-manager/shared/tsconfig.json should match snapshot 1`] = `
"{
  "extends": "../../../tsconfig.base.json",
  "compilerOptions": {
    "paths": {
      "@testcompany/task-manager-shared": [
        "libs/task-manager/shared/src/index.ts"
      ],
      "@": [
        "libs/task-manager/shared/src"
      ],
      "@/*": [
        "libs/task-manager/shared/src/*"
      ],
      "@test/*": [
        "libs/task-manager/shared/test/helper/*"
      ]
    }
  },
  "files": [],
  "include": [],
  "references": [
    {
      "path": "./tsconfig.test.json"
    },
    {
      "path": "./tsconfig.build.json"
    }
  ]
}"
`;

exports[`Library Generator: Shared libs/task-manager/shared/tsconfig.test.json should match snapshot 1`] = `
"{
  "extends": "./tsconfig.json",
  "compilerOptions": {
    "outDir": "../../../dist/out-tsc",
    "types": ["jest", "node"]
  },

  "files": [
    "../../../node_modules/jest-expect-message/types/index.d.ts",
    "../../../node_modules/@hexancore/common/lib/cjs/Test/Matchers.d.ts"
  ],
  "include": [
    "src/**/*.ts",
    "test/**/*.test.ts"
  ]
}
"
`;

exports[`Library Generator: Shared nx.json should match snapshot 1`] = `"{"affected":{"defaultBase":"main"},"targetDefaults":{"build":{"cache":true},"lint":{"cache":true}}}"`;

exports[`Library Generator: Shared package.json should match snapshot 1`] = `
"{
  "name": "@testcompany/source"
}
"
`;

exports[`Library Generator: Shared tsconfig.base.json should match snapshot 1`] = `
"{
  "compilerOptions": {
    "paths": {
      "@testcompany/task-manager-shared": [
        "libs/task-manager/shared/src/index.ts"
      ]
    }
  }
}
"
`;
